build.gradle

```diff
@@ -279,6 +295,78 @@ subprojects {
       exceptionFormat = testExceptionFormat
     }
     logTestStdout.rehydrate(delegate, owner, this)()
+
+    afterTest { TestDescriptor desc , TestResult result ->
+      if (TestResult.ResultType.FAILURE == result.resultType) {
+        logger.warn("Found failing test $desc.parent - $desc.name, trying again in next turn of test execution")
+        failedTestsInFirstTestRun << desc
+      }
+    }
+    testResultsDir = reports.junitXml.destination
+
+    ignoreFailures = shouldFilterFlakyTests
+  }
+
+  task testWithFlakyRetry(type: Test, dependsOn: test) {
+    maxParallelForks = userMaxForks ?: Runtime.runtime.availableProcessors()
+
+    minHeapSize = "256m"
+    maxHeapSize = "2048m"
+
+    testLogging {
+      events = userTestLoggingEvents ?: testLoggingEvents
+      showStandardStreams = userShowStandardStreams ?: testShowStandardStreams
+      exceptionFormat = testExceptionFormat
+    }
+
+    def oldResultsToOverwite = []
+
+    doFirst {
+      if (!failedTestsInFirstTestRun.isEmpty()) {
+        logger.warn("Executing failed tests again to filter out flaky tests")
+        logger.warn("Failed tests: " + failedTestsInFirstTestRun)
+
+        logger.warn("Successful results will overwrite stale failed test results from $testResultsDir to avoid jenkins reporting an UNSTABLE build")
+        def failedTestClasses = failedTestsInFirstTestRun.collect{
+          it.className.replaceAll('\\.', '/') + ".*"
+        } as Set
+        // re-running complete classes to keep the integrity of the test results
+        logger.warn("Re-running $failedTestClasses")
+        setIncludes(failedTestClasses)
+      } else {
+        logger.info("Yay, no failing tests, no need for a second run!")
+        setExcludes(["**/*"])
+      }
+    }
+
+    afterSuite { TestDescriptor desc , TestResult result ->
+      if (TestResult.ResultType.FAILURE != result.resultType) {
+        // In case of flaky tests old results should be overwritten to avoid Jenkins considering a successful build
+        // UNSTABLE due to failed tests
+        // Unfortunately at this point new result files are not ready, so we need to defer overwrites to a later time,
+        // see doLast below
+        def testClassName = desc.className
+        // Gradle seems to call afterSuite not only after suites but also when the executor finishes
+        // no need to overwrite anything in this case
+        if (testClassName != null) {
+          logger.info("Second execution of $testClassName was successful, previous failures seem to be flaky. Marking test result to be overwritten.")
```

Comment:
Wouldn't this get logged after every test class ran, not only the failed ones?

Reply:
No, it would only run after the failed ones, so this is how it would look like:
```
> Task :clients:testWithFlakyRetry
Task ':clients:testWithFlakyRetry' is not up-to-date because:
  Task has failed previously.
Executing failed tests again to filter out flaky tests
Failed tests: [Test flakyTest(org.apache.kafka.clients.ApiVersionsTest)]
Successful results will overwrite stale failed test results from /Users/viktor.somogyi/dev/github/apache/kafka/clients/build/test-results/test to avoid jenkins reporting an UNSTABLE build
Re-running 1 failing test classes
Starting process 'Gradle Test Executor 44'. Working directory: /Users/viktor.somogyi/dev/github/apache/kafka/clients Command: /Library/Java/JavaVirtualMachines/jdk-11.0.1.jdk/Contents/Home/bin/java -Dorg.gradle.native=false -javaagent:build/tmp/expandedArchives/org.jacoco.agent-0.8.2.jar_2aca8b620b19ecd063f63feff8caaa38/jacocoagent.jar=destfile=build/jacoco/testWithFlakyRetry.exec,append=true,inclnolocationclasses=false,dumponexit=true,output=file,jmx=false @/private/var/folders/km/99nb__7s2w7ftd7jbvs9m0mm0000gp/T/gradle-worker-classpath662232928963633522txt -Xms256m -Xmx2048m -Dfile.encoding=UTF-8 -Duser.country=HU -Duser.language=en -Duser.variant -ea worker.org.gradle.process.internal.worker.GradleWorkerMain 'Gradle Test Executor 44'
Successfully started process 'Gradle Test Executor 44'
Second execution of org.apache.kafka.clients.ApiVersionsTest was successful, previous failures seem to be flaky. Marking test result to be overwritten.
Finished generating test XML results (0.0 secs) into: /Users/viktor.somogyi/dev/github/apache/kafka/clients/build/test-results/testWithFlakyRetry
Generating HTML test report...
Finished generating test html results (0.001 secs) into: /Users/viktor.somogyi/dev/github/apache/kafka/clients/build/reports/tests/testWithFlakyRetry
Overwriting old test result at /Users/viktor.somogyi/dev/github/apache/kafka/clients/build/test-results/test/TEST-org.apache.kafka.clients.ApiVersionsTest.xml with /Users/viktor.somogyi/dev/github/apache/kafka/clients/build/test-results/testWithFlakyRetry/TEST-org.apache.kafka.clients.ApiVersionsTest.xml
:clients:testWithFlakyRetry (Thread[Execution worker for ':' Thread 4,5,main]) completed. Took 1.831 secs.

Deprecated Gradle features were used in this build, making it incompatible with Gradle 6.0.
Use '--warning-mode all' to show the individual deprecation warnings.
See https://docs.gradle.org/5.0/userguide/command_line_interface.html#sec:command_line_warnings

BUILD SUCCESSFUL in 1m 51s
13 actionable tasks: 6 executed, 7 up-to-date
```